atom/atom
# Update find-and-replace

### Requirements for Contributing a Bug Fix

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix an existing bug. To contribute other changes, you must use a different template. You can see all templates at https://github.com/atom/atom/tree/master/.github/PULL_REQUEST_TEMPLATE.
* The pull request must update the test suite to demonstrate the changed functionality. For guidance, please see https://flight-manual.atom.io/hacking-atom/sections/writing-specs/.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see https://github.com/atom/atom/tree/master/CONTRIBUTING.md#pull-requests.

### Identify the Bug

See https://github.com/atom/find-and-replace/pull/966, which fixes 5 issues!

find-and-replace was mistakenly assuming that any object with a buffer was also a text editor, which is usually but not always the case, leading to some nasty uncaught exceptions when using [plantuml-viewer](https://atom.io/packages/plantuml-viewer).

### Description of the Change

@howyi and I worked together to craft a PR that makes sure the object is a TextEditor instead of checking for the existence of a buffer.

### Alternate Designs

There were a couple of alternatives explored - accessing the buffer using `getBuffer`, also making sure the `buffer` property exists - but none of them fixed the root cause.

### Possible Drawbacks

In the _very_ unlikely event that there's some Atom package out there that implements all the methods that find-and-replace needs to perform correctly yet isn't a TextEditor, find-and-replace will no longer work with that package.

### Verification Process

1. Install plantuml-viewer (and Java)
2. Open a plantuml viewer
3. In the editor that belongs to it, activate find-and-replace
4. Focus the plantuml viewer and verify that no exceptions are thrown

The test suite was also run.

### Release Notes

Fixed an issue where find-and-replace would not work correctly with plantuml-viewer.
